CMakeLists.txt

```diff
@@ -557,6 +564,41 @@ endif()
 
 include(CMakePackageConfigHelpers)
 
+# If we install a dynamically linked binary that also uses dynamically used
+# libraries from deps/install/lib, we need to be cautious. Changing these
+# shared libraries from deps/install/lib most probably breaks the binary.
+# We only allow such an installation for custom installation prefixes
+# (in the assumption that only reasonably experienced users use this and 
+# also that custom installation prefixes are not used for longer periods of
+# time anyway). Also, we print a big warning with further instructions.
+if(NOT ENABLE_STATIC_BINARY)
+  # Get the libraries that cvc4 links against
+  get_target_property(libs cvc4 INTERFACE_LINK_LIBRARIES)
+  set(LIBS_SHARED_FROM_DEPS "")
+  foreach(lib ${libs})
+    # Filter out those that are linked dynamically and come from deps/install
+    if(lib MATCHES ".*/deps/install/lib/.*\.so")
+      list(APPEND LIBS_SHARED_FROM_DEPS ${lib})
+    endif()
+  endforeach()
+  list(LENGTH LIBS_SHARED_FROM_DEPS list_len)
+  # Check if we actually use such "dangerous" libraries
+  if(list_len GREATER 0)
+    # Print a generic warning
+    install(CODE "message(WARNING \"You are installing a dynamically linked binary of CVC4 which may be a problem if you are using any dynamically linked third-party library that you obtained through one of the contrib/get-xxx scripts. The binary uses the rpath mechanism to find these locally, hence executing such a contrib script removing the \\\"deps/install\\\" folder most probably breaks the installed binary! Consider installing the dynamically linked dependencies on your system manually or link cvc4 statically.\")")
+    # Print the libraries in question
+    foreach(lib ${LIBS_SHARED_FROM_DEPS})
+      install(CODE "message(WARNING \"The following library is used by the cvc4 binary: ${lib}\")")
+    endforeach()
+    # Check if we use a custom installation prefix
+    if(CMAKE_INSTALL_PREFIX EQUAL "/usr/local")
```

Comment:
Use `if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)` instead of checking `/usr/local`.

 https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT.html#variable:CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT

Reply:
This is an interesting issue I already pondered when implementing this. The issue is, that we want to check something related, but different to what `CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` tells us.
`CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` is true if there was no `-DCMAKE_INSTALL_PREFIX=...` passed to cmake. However, if it was passed, this does not mean everything is fine. A user (or another script) could pass `/usr/local` explicitly. A user could change the build type with `ccmake`. In both these cases, simply checking `CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` is not sufficient, but the current check is fine.

I guess I don't see good solutions here. For example, we would not warn users installing to `/usr` or `/opt` or alike. But I still prefer the current one over `CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` as it warns slightly more users that are at risk.